neuralpredictors/layers/cores/conv2d.py

```diff
@@ -182,8 +182,11 @@ def add_bn_layer(self, layer):
                     layer["scale"] = self.scale_layer_cls(self.hidden_channels)
 
     def add_activation(self, layer):
-        if (self.num_layers > 1 or self.final_nonlinearity) and not self.linear:
-            layer["nonlin"] = AdaptiveELU(self.elu_xshift, self.elu_yshift)
+        if self.linear:
+            return
+        if len(self.features) == self.num_layers - 1 and not self.final_nonlinearity:
+            return
+        layer["nonlin"] = AdaptiveELU(self.elu_xshift, self.elu_yshift)
```

Comment:
I would personally prefer to make the condition more explicit about when a nonlinearity will be added:
```python
if self.linear:
    return
if len(self.features) < self.num_layers - 1 or self.final_nonlinearity:
    layer["nonlin"] = AdaptiveELU(self.elu_xshift, self.elu_yshift)
```

Reply:
I changed that.